src/pseudopeople/noise_functions.py

```diff
@@ -608,37 +585,140 @@ def make_ocr_errors(
-    token_noise_level = configuration[Keys.TOKEN_PROBABILITY]
-    rng = np.random.default_rng(seed=get_hash(f"{randomness_stream.seed}_make_ocr_errors"))
-    column = data[column_name]
-    column = column.astype(str)
-    for idx in column.index:
-        noised_value = ocr_corrupt(
-            column.loc[idx],
-            token_noise_level,
-            rng,
-        )
-        column[idx] = noised_value
+    ocr_errors = load_ocr_errors()
 
-    return column
+    return _corrupt_tokens(
+        ocr_errors,
+        data[column_name].astype(str),
+        configuration[Keys.TOKEN_PROBABILITY],
+        randomness_stream,
+        addl_key="make_ocr_errors",
+    )
+
+
+def _corrupt_tokens(
+    errors: pd.DataFrame,
+    column: pd.Series,
+    token_probability: float,
+    randomness_stream: RandomnessStream,
+    addl_key: str,
+) -> pd.Series:
+    """
+    Performs token-level corruption on a string Series when the tokens to corrupt
+    (and the tokens they get corrupted to) can be more than one character long.
+    A token does not need to be the same length as a token it gets corrupted to.
+    When both kinds of tokens are always one character, things are simpler,
+    and a faster algorithm can be used; see typographic noise for an example of this.
+
+    `errors` is a pandas DataFrame where the index contains tokens that are eligible to
+    be corrupted, and the columns contain the corruptions that are possible.
+    The column names do not matter.
+    """
+    # NOTE: This first section depends only on `errors` and could be cached -- we might
+    # consider moving it into the load_* functions.
+    # However, the amount of work here scales only with the size of `errors` and not
+    # with the data.
+    # It isn't possible to just make this a cached helper since `errors` is a DataFrame,
+    # which is unhashable.
+    max_token_length = errors.index.str.len().max()
+    errors_eligible_tokens = (
+        pd.Series(errors.index).groupby(errors.index.str.len()).apply(set).to_dict()
+    )
+    number_of_options = errors.count(axis=1)
+    errors_stacked = errors.stack()
+
+    lengths = column.str.len().values
+
+    same_len_col_exploded = (
+        column
+        # Convert to numpy string dtype
+        .values.astype(str)
+        .view("U1")
+        .reshape((len(column), lengths.max()))
+    )
+
+    rng = np.random.default_rng(seed=get_hash(f"{randomness_stream.seed}_{addl_key}"))
+
+    # NOTE: Somewhat surprisingly, this seemed to perform better using Python string types than NumPy types.
+    # Perhaps worth more investigation in the future.
+    # Note that each item in result can be up to the number of characters of the *longest* corrupted token.
+    result = np.empty(same_len_col_exploded.shape, dtype=str).astype("object")
+    # Tracks the next character index where we have to pay attention to a given string.
+    # When we corrupt a token, we never corrupt again within that same source token, so we
```

Comment:
Reword: "When we corrupt a token, we skip to the end of that corrupted token before corrupting anything new"